x/pairing/keeper/pairing.go

```diff
@@ -137,16 +140,42 @@ func (k Keeper) getPairingForClient(ctx sdk.Context, chainID string, clientAddre
 		return nil, 0, "", err
 	}
 
-	providers, err = k.calculatePairingForClient(ctx, possibleProviders, project.Index, block,
-		chainID, epochHash, strictestPolicy.MaxProvidersToPair)
+	// create the pairing slots with assigned reqs
+	slots := pairingscores.CalcSlots(strictestPolicy, minStake)
+
+	// group identical slots (in terms of reqs types)
+	slotGroups := pairingscores.GroupSlots(slots)
+
+	// create providerScore array with all possible providers
+	providerScores := []*pairingscores.PairingScore{}
+	for i := range possibleProviders {
+		providerScore := pairingscores.NewPairingScore(&possibleProviders[i])
+		providerScores = append(providerScores, providerScore)
+	}
+
+	// calculate score (always on the diff in score components of consecutive groups) and pick providers
+	prevGroupSlot := pairingscores.NewPairingSlot(map[string]pairingscores.ScoreReq{}) // init dummy slot to compare to
+	indexToSkipMapPtr := make(map[int]bool)                                            // keep the indices of chosen providers to we won't pick the same providers twice (for different groups)
+	for _, group := range slotGroups {
+		diffSlot := group.Slot.Diff(prevGroupSlot)
```

Comment:
What is the reason/purpose of using the diff here?

Reply:
When I'm calculating the score for a slot, I can calculate only parts of the score if the previous and current slots have shared score components. For example, let's say one slot has stake+geo-eu+archive and the next one has stake+geo-eu. Instead of calculating the next slot's score from scratch I can just remove the archive score component from the previous score

Reply:
it would require careful handling in that case.
but anyhow I see this code is gone.